From f9d943a1d925d079a0b66ff017c2fe374e9e5c93 Mon Sep 17 00:00:00 2001
From: Ken Rockot <rockot@google.com>
Date: Wed, 20 Nov 2024 23:22:48 +0000
Subject: [PATCH] ipcz: Validate consistent subparcel counts

If a parcel is decomposed into subparcels for piecemeal transmission,
each subparcel must agree on the number of subparcels which comprise the
full parcel. This agreement was not being validated properly.

Fixed: 379207974
Change-Id: I2b1d1193effd59043ef1b1af3bd41ebceb907b99
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6035939
Reviewed-by: Alex Gough <ajgo@chromium.org>
Commit-Queue: Ken Rockot <rockot@google.com>
Cr-Commit-Position: refs/heads/main@{#1385945}
---

--- a/third_party/ipcz/src/ipcz/node_link.cc
+++ b/third_party/ipcz/src/ipcz/node_link.cc
@@ -968,6 +968,10 @@
     SubparcelTracker& tracker = it->second;
     if (inserted) {
       tracker.subparcels.resize(num_subparcels);
+    } else if (tracker.subparcels.size() != num_subparcels) {
+      // Inconsistent subparcel count expectations across subparcels. This is
+      // a validation failure.
+      return false;
     }
 
     // Note that `index` has already been validated against the expected number
